Src/PerformanceCollector/Shared/Implementation/QuickPulse/Process/QuickPulseTopCpuCollector.cs

```diff
@@ -0,0 +1,154 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.PerfCounterCollector.Implementation.QuickPulse
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Security;
+
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+    using Microsoft.ApplicationInsights.Extensibility.PerfCounterCollector.Implementation.QuickPulse.Helpers;
+
+    /// <summary>
+    /// Top CPU collector.
+    /// </summary>
+    internal sealed class QuickPulseTopCpuCollector : IQuickPulseTopCpuCollector
+    {
+        private readonly TimeSpan accessDeniedRetryInterval = TimeSpan.FromMinutes(1);
+
+        private readonly Clock timeProvider;
+
+        private readonly IQuickPulseProcessProvider processProvider;
+
+        // process name => (last observation timestamp, last observation value)
+        private readonly Dictionary<string, TimeSpan> processObservations = new Dictionary<string, TimeSpan>(StringComparer.Ordinal);
+
+        private DateTimeOffset prevObservationTime;
+
+        private TimeSpan? prevOverallTime;
+
+        private DateTimeOffset lastReadAttempt = DateTimeOffset.MinValue;
+
+        public QuickPulseTopCpuCollector(Clock timeProvider, IQuickPulseProcessProvider processProvider)
+        {
+            this.timeProvider = timeProvider;
+            this.processProvider = processProvider;
+
+            this.InitializationFailed = false;
+            this.AccessDenied = false;
+        }
+
+        public bool InitializationFailed { get; private set; }
+
+        public bool AccessDenied { get; private set; }
+
+        public IEnumerable<Tuple<string, int>> GetTopProcessesByCpu(int topN)
+        {
+            try
+            {
+                DateTimeOffset now = this.timeProvider.UtcNow;
+
+                if (this.InitializationFailed)
+                {
+                    return Enumerable.Empty<Tuple<string, int>>();
+                }
+
+                if (this.AccessDenied && now - this.lastReadAttempt < this.accessDeniedRetryInterval)
+                {
+                    return Enumerable.Empty<Tuple<string, int>>();
+                }
+
+                var procData = new List<Tuple<string, double>>();
+                var encounteredProcs = new HashSet<string>();
+                
+                this.lastReadAttempt = now;
+
+                TimeSpan? totalTime;
+                foreach (var process in this.processProvider.GetProcesses(out totalTime))
+                {
+                    encounteredProcs.Add(process.ProcessName);
+                    
+                    TimeSpan lastObservation;
+                    if (!this.processObservations.TryGetValue(process.ProcessName, out lastObservation))
+                    {
+                        // this is the first time we're encountering this process
+                        this.processObservations.Add(process.ProcessName, process.TotalProcessorTime);
+
+                        continue;
```

Comment:
We are capturing the time (note process.TotalProcessorTime usage in this line). We can't do much more (since we need two data points to calculate the CPU usage), so we just add the new process to the collection - along with the original value of its time value.